### ISSUE TITLE
Memory Leak on GroupedStream.filter

### ISSUE THREAD

**Issue by ZeLonewolf** [reactor/reactor#572]
I am seeing a massive memory leak when using GroupedStream.filter() on reactor 2.0.7.  I also saw the issue on reactor 2.0.2, which I was using before upgrading in an attempt to solve the problem.  Using a profiler, many instances of java.util.concurrent.LinkedTransferQueue$Node are created but never garbage collected.  I was able to make a short reproducer that demonstrates this behavior.  It goes out of memory on a 1GB JVM in approximately 45 seconds on my machine.
````java
import java.util.stream.LongStream;

import reactor.Environment;
import reactor.rx.broadcast.Broadcaster;

public class Test
{
    private static Broadcaster<Long> bcast = Broadcaster.<Long> create();

    public static void main(String... args)
    {
        Environment.initializeIfEmpty();

        bcast.groupBy(i -> i % 10).filter(gs -> gs.key().equals(5L))
                .consume(stream -> stream.consume(System.out::println));

        LongStream.range(0, Long.MAX_VALUE).forEach(bcast::onNext);
    }
}
````

**Comment by smaldini** [comment]
Thanks for the detailled report, it seems the issue is fixed on master based on the new groupBy engine (bounded, so no risk of OoM, and which will be soon deployed on a new 2.1 snapshot after the CI is fixed). However I'll see if there is an alternative fix for this issue.

**Comment by ZeLonewolf** [comment]
Great, looking forward to the update.  Below is the workaround I've come up with to achieve the same behavior.  Ugly and verbose but it will make groupBy work until the next version by doing the inverse check of the filter predicate and closing the non-matching streams.

````java
package com.example;

import java.util.stream.LongStream;

import reactor.Environment;
import reactor.fn.Predicate;
import reactor.rx.broadcast.Broadcaster;
import reactor.rx.stream.GroupedStream;

public class Test {
	private static Broadcaster<Long> bcast = Broadcaster.<Long> create();

	public static void main(String... args) {
		Environment.initializeIfEmpty();

		Predicate<GroupedStream<Long, Long>> five = gs -> gs.key().equals(5L);

		bcast.groupBy(i -> i % 10).observe(gs -> {
			if (!five.test(gs)) {
				gs.consume().cancel();
			}
		}).filter(five).consume(stream -> stream.consume(System.out::println));

		LongStream.range(0, Long.MAX_VALUE).forEach(bcast::onNext);
	}
}
````

**Comment by smaldini** [comment]
Should be fixed in reactor-stream 2.5.0.M1 [groupBy](http://next.projectreactor.io/stream/docs/api/reactor/rx/Stream.html#groupBy-reactor.fn.Function-) .
